File: Python/647. Palindromic Substrings.py
class Solution(object):
    def countSubstrings(self, s):
        """
        :type s: str
        :rtype: int
        """
        n = len(s)
        dp = [[0]*n for i in range(n)]
        ans = 0
        for i in range(n):
            for j in range(i+1):
                if s[i] == s[j]:
                    if i-j < 2:
                        dp[j][i] = 1
                    else:
                        dp[j][i] = dp[j+1][i-1]
                if dp[j][i]:
                    ans += 1
        return ans

File: C++/647. Palindromic Substrings.cpp
class Solution {
public:
    int countSubstrings(string s) {
        int len = s.length();
        int ans = 0;
        vector<vector<int> > dp(len, vector<int>(len, 0));
        for(int j=0;j<s.length();j++)
        {
            for(int i=0;i<=j;i++)
            {
                if(s[i]==s[j])
                {
                    if(j - i < 2)
                        dp[i][j] = 1;
                    else
                        dp[i][j] = dp[i+1][j-1];
                }
                if(dp[i][j])
                    ans++;
            }
        }
        return ans;
    }
};